Update from Forestry.io Alex Potsides created src/_blog/js-ipfs-0.55.0-greatly-improves-type-definitions.md (#191)
ipfs/ipfs-blog: src/_blog/js-ipfs-0.55.0-greatly-improves-type-definitions.md

---
title: js-IPFS 0.55.0 greatly improves type definitions
description: The new release comes with types rewritten from the ground up
author: Alex Potsides
date: 2021-05-10
permalink: "/2021-05-10-js-ipfs-0-55/"
translationKey: ''
header_image: "/header-image-js-ipfs-placeholder.png"
tags:
- js-ipfs

---
# 🔦 Highlights

> > TypeScript types have been rewritten from the ground up

Reviewer: ```suggestion
> TypeScript type definitions have been rewritten from the ground up
```

## 🍪 Types

It's been a long time in the works, but `js-ipfs@0.55.0` finally lands with a brand new set of TypeScript type definitions entirely re-written from the ground up.

> We previously shipped `0.51.0` with bundled TypeScript type definitions which enabled users to explore the IPFS Core API through [intellisense](https://code.visualstudio.com/docs/editor/intellisense) and use that to ensure they were calling methods that existed, but the argument and return types were frequently marked `any` which although would not cause TypeScript compilation errors, gave you little in the way of actual type safety.

Reviewer 2: ```suggestion
We previously shipped `0.51.0` with bundled TypeScript type definitions which enabled users to explore the IPFS Core API through [intellisense](https://code.visualstudio.com/docs/editor/intellisense) and use that to ensure they were calling methods that existed, but the argument and return types were frequently marked `any`, which, although would not cause TypeScript compilation errors, gave you little in the way of actual type safety.
```

Part of the problem is that `js-ipfs` exposes the return types of many supporting modules which do not ship with TypeScript definitions.  This means we either had to PR support for TypeScript definitions into these modules or type their input/output at the `js-ipfs` level which is error prone and makes no guarantees about the actual underlying code.

> We took a look at our stack and started to add TypeScript types from the lowest level up, which has been an enormous job with some 250 pull requests opened, reviewed, merged and shipped as part of this effort, massive thanks to everyone involved, see the contributors list further down this post for who helped out!

Reviewer 2: ```suggestion
We took a look at our stack and started to add TypeScript types from the lowest level up, which has been an enormous job with some 250 pull requests opened, reviewed, merged and shipped as part of this effort. A massive thanks to everyone involved, see the contributors list further down this post for who helped out!
```

## 🙅 noImplicitAny

Early on we decided to enable [noImplicitAny](https://www.typescriptlang.org/tsconfig) in our `tsconfig.json` files.  This is a strict setting which causes an error when type information for any variable cannot be found.

It means that everything has to be typed internally which increased the amount of work necessary to deliver this release but our internal code is now vastly safer and it even surfaced a few bugs and unhandled edge cases in the implementation so upgrading is very much recommended.

## 🤖 Future proofing

> In line with our [supported versions](https://github.com/ipfs/community/blob/master/CONTRIBUTING_JS.md#supported-versions), `js-ipfs@0.55.0` has dropped support for node < 14, this is so we can support the latest and greatest features without having to carry legacy baggage forward.

Reviewer 2: ```suggestion
In line with our [supported versions](https://github.com/ipfs/community/blob/master/CONTRIBUTING_JS.md#supported-versions), `js-ipfs@0.55.0` has dropped support for Node.js < 14. This is so we can support the latest and greatest features without having to carry legacy baggage forward.
```

> We've also switched to using named exports for our top-level modules instead of default exports as it makes them more pleasant to consume from [ES6 module](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) environments.  It also means the type defs generated from [JSDoc comments](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) are more compact and have to jump through fewer hoops to reflect the code they are generated from.

Reviewer 2: ```suggestion
We've also switched to using named exports for our top-level modules instead of default exports as it makes them more pleasant to consume from [ES module](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) environments.  It also means the type definitions generated from [JSDoc annotations](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) are more compact and have to jump through fewer hoops to reflect the code they are generated from.
```

> Code wise this only affects ipfs-http-client:

Reviewer 2: ```suggestion
From perspective of the external API, this only affects ipfs-http-client:
```

```js
// before
const create = require('ipfs-http-client')
const ipfs = create()

// after
const { create } = require('ipfs-http-client')
const ipfs = create()
```

> Finally in some places we return instances of the [bignumber.js](https://www.npmjs.com/package/bignumber.js) module - this has been necessary in the past because JavaScript lacked an arbitrary precision number type.  [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) has been present in all of our supported environments for some time so we've been removing `bignumber.js` in favour of `BigInt` in the [Core API](https://github.com/ipfs/js-ipfs/tree/master/docs).

Reviewer 2: ```suggestion
Finally, in some places we previously returned instances of the [bignumber.js](https://www.npmjs.com/package/bignumber.js) module—this has been necessary in the past because JavaScript lacked an arbitrary precision number type.  [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) has been present in all of our supported environments for some time so we've been removing `bignumber.js` in favour of `BigInt` in the [Core API](https://github.com/ipfs/js-ipfs/tree/master/docs).
```

Unclear to me if we've done this for all instances of bignumber.js or just some? I know borc still depends on it but it should rarely get involved, if at all for users. I think this suggested change phrases it more as "we've removed it all" but the last part about Core API suggests maybe it's not a complete work? Maybe clarify if it's not complete and what the path might be there.

Author: It's been removed from the core api, but until we do the multiformats upgrade a cbor object could have `bignumber.js` since that's how `borc` does them as you point out.  I figure that's not part of the core api?

> ## 🛍️ Upgrade Guide

Reviewer 2: ```suggestion
# 🛍️ Upgrade Guide
```

> We've taken this opportunity to align the implementation with the published [Core API Docs](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api) - in some cases the accepted input types were broader than what was documented for backwards compatibility, but that compatibility comes at the cost of code complexity and added maintenance, so those old code paths have been removed.

Reviewer 2: ```suggestion
We've taken this opportunity to align the implementation with the published [Core API Docs](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api). In some cases the accepted input types were broader than what was documented for backward compatibility, but that compatibility comes at the cost of code complexity and added maintenance, so those old code paths have been removed.
```

> We've taken this opportunity to align the implementation with the published [Core API Docs](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api) - in some cases the accepted input types were broader than what was documented for backwards compatibility, but that compatibility comes at the cost of code complexity and added maintenance, so those old code paths have been removed.

Reviewer: Can we now recommend the built in docs created? Or do we still need to get them improved? (We currently have the docs as markdown docs in the js-ipfs repo, but it would be great if we do not need to maintain them and guarantee they are updated)

Author: I was going to do that, but I don't think aegir's doc command plays well with monorepos so it may need a bit more thought - cc @hugomrdias

If you have coded against the Core API docs you should have very few surprises in store.

The API changes are as follows:

* `ipfs.bitswap.stat()` - returned `peers` is a `string[]` (was `CID[]`)
* `ipfs.bitstwap.stat()` - returned `blocksReceived` is a `BigInt` (was `bignumber.js`)
* `ipfs.bitstwap.stat()` - returned `dataReceived` is a `BigInt` (was `bignumber.js`)
* `ipfs.bitstwap.stat()` - returned `blocksSent` is a `BigInt` (was `bignumber.js`)
* `ipfs.files.rm(path, opts)` - `path` arg is `string | string[]` (was `...string`)
* `ipfs.files.cp(source, dest, opts)` - `source` arg is `string | string[]` (was `...string`)
* `ipfs.files.mv(source, dest, opts)` - `source` arg is `string | string[]` (was `...string`)

> # ✨New features

Reviewer 2: ```suggestion
## ✨New features
```

not quite sure what to do with the H2 and H1 structure here - New features was H1 but Breaking changes and Bug fixes are H2 below it. Suggesting that these are all H2's under Upgrade Guide, but also they could be under a new H1 called "Changes" which might fit better.

> * support identity hash in block.get + dag.get ([#3616](https://github.com/ipfs/js-ipfs/issues/3616)) ([28ad9ad](https://github.com/ipfs/js-ipfs/commit/28ad9ad6e50abb89a366ecd6b5301e848f0e9962))

Reviewer 2: ```suggestion
* Support identity hash ([`0x00](https://github.com/multiformats/multicodec/blob/master/table.csv#L2)) in `ipfs.block.get()` + `ipfs.dag.get()` ([#3616](https://github.com/ipfs/js-ipfs/issues/3616)) ([28ad9ad](https://github.com/ipfs/js-ipfs/commit/28ad9ad6e50abb89a366ecd6b5301e848f0e9962))
```

## 🔨 Breaking changes

* The minimum supported Node.js version is 14

> * All core api methods now have types, some method signatures have changed (see Upgrade guide, above)

Reviewer 2: ```suggestion
* All Core API methods now have types, some method signatures have changed (see Upgrade Guide above)
```

* ignore the ts error caused by the recent protobufjs type change ([#3656](https://github.com/ipfs/js-ipfs/issues/3656)) ([084589c](https://github.com/ipfs/js-ipfs/commit/084589c0116d8f27ce1462424fb93b6037b776a9))
* update data type for ws message event handler ([#3641](https://github.com/ipfs/js-ipfs/issues/3641)) ([4a14d20](https://github.com/ipfs/js-ipfs/commit/4a14d20e727b50a8d98c14573d9a5b6fa0e8699d))
* do not republish self key twice ([#3634](https://github.com/ipfs/js-ipfs/issues/3634)) ([8545a76](https://github.com/ipfs/js-ipfs/commit/8545a763daa38aefa71cca514016ba400363830a))
* fix types ([#3662](https://github.com/ipfs/js-ipfs/issues/3662)) ([0fe8892](https://github.com/ipfs/js-ipfs/commit/0fe8892361180dab53ed3c3b006479b32a792d44))
* update ipfs repo ([#3671](https://github.com/ipfs/js-ipfs/issues/3671)) ([9029ee5](https://github.com/ipfs/js-ipfs/commit/9029ee591fa74ea65c9600f2d249897e933416fa))
* update types after feedback from ceramic ([#3657](https://github.com/ipfs/js-ipfs/issues/3657)) ([0ddbb1b](https://github.com/ipfs/js-ipfs/commit/0ddbb1b1deb4e40dac3e365d7f98a5f174c2ce8f)), closes [#3640](https://github.com/ipfs/js-ipfs/issues/3640)

> # 🗺️ What’s next?

Reviewer 2: ```suggestion
# 🗺️ What’s Next?
```

:shrug: just trying to make capitalisation consistent but it breaks down a bit with the next couple of H1's, but they're sentences so maybe that's ok

Author: It would need doing to "Breaking changes", "Bug fixes", etc too I guess.  Maybe "Upgrade Guide" just needs to be "Upgrade guide" instead.

* [@valmack](https://github.com/valmack) (1 issue, 3 comments)
* [@vasco-santos](https://github.com/vasco-santos) (230 commits, 66 PRs, 8 issues, 167 comments)
* [@vmx](https://github.com/vmx) (41 commits, 11 PRs, 3 issues, 31 comments)
* [@vogdb](https://github.com/vogdb) (1 commit, 9 comments)
* [@wehriam](https://github.com/wehriam) (1 issue, 1 comment)
* [@wemeetagain](https://github.com/wemeetagain) (1 commit, 1 PR, 1 issue, 2 comments)
* [@xmaysonnave](https://github.com/xmaysonnave) (2 issues, 4 comments)
* [@xzy-96](https://github.com/xzy-96) (1 issue)
* [@yuhr](https://github.com/yuhr) (1 issue, 3 comments)
* [@yurtsiv](https://github.com/yurtsiv) (1 commit, 2 PRs, 1 issue, 10 comments)
* [@yusefnapora](https://github.com/yusefnapora) (1 PR, 4 comments)
* [@zerdos](https://github.com/zerdos) (1 issue)
* [@Zorlin](https://github.com/Zorlin) (1 issue, 3 comments)

# 🙌🏽 Want to contribute?

Would you like to contribute to the IPFS project and don’t know how? Well, there are a few places you can get started:

* Check the issues with the `help wanted` label in the [js-IPFS repo](https://github.com/ipfs/js-ipfs/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)

> * Join an IPFS All Hands, introduce yourself and let us know where you would like to contribute: [https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands](https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands "https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands")

Reviewer 2: just a thought, you could also add the IPLD call which would be pretty relevant for people wanting to talk about data, maybe:

```suggestion
* Join an IPFS All Hands, introduce yourself and let us know where you would like to contribute: [https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands](https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands "https://github.com/ipfs/team-mgmt/#weekly-ipfs-all-hands")
* Join an IPLD All Hands call if you're interested in the data layer of IPFS: [https://github.com/ipld/team-mgmt#every-two-weeks-call](https://github.com/ipld/team-mgmt#every-two-weeks-call)
```

Reviewer: this probably comes from a template? This should also be updated there

Author: Just copy/pasted from the previous js-IPFS blog post tbh..

> * Hack with IPFS and show us what you made! The All Hands call is also the perfect venue for demos, join in and show us what you built

Reviewer 2: ```suggestion
* Hack with IPFS and show us what you made! The All Hands calls are also the perfect venue for demos, join in and show us what you built
```

if IPLD is added

Reviewer: this probably comes from a template? This should also be updated there
